# `boolean-naming-convention`

Enforces boolean variables, parameters, properties, and functions to use consistent prefixes such as **is**, **has**, **can**, **should**, etc.  
Consistent prefixes improve readability and make boolean intent explicit.

---

## Table of contents

- [`boolean-naming-convention`](#boolean-naming-convention)
  - [Table of contents](#table-of-contents)
  - [Rule summary](#rule-summary)
  - [What the rule checks](#what-the-rule-checks)
  - [Options (all configurations)](#options-all-configurations)
    - [Default options](#default-options)
    - [Option details](#option-details)
      - [allowedPrefixes](#allowedprefixes)
      - [ignore](#ignore)
      - [checkFunctions](#checkfunctions)
      - [checkParameters](#checkparameters)
      - [checkProperties](#checkproperties)
      - [checkVariables](#checkvariables)
  - [Examples (by option)](#examples-by-option)
    - [Default behavior](#default-behavior)
    - [Custom prefixes](#custom-prefixes)
    - [Disabling specific checks](#disabling-specific-checks)
      - [Example](#example)
    - [Destructured params \& typed object patterns](#destructured-params--typed-object-patterns)
    - [Function-valued properties and variable initializers](#function-valued-properties-and-variable-initializers)
    - [Destructured variable declarators (typed)](#destructured-variable-declarators-typed)
  - [Messages](#messages)
  - [Implementation notes \& requirements](#implementation-notes--requirements)
  - [Limitations \& edge cases](#limitations--edge-cases)
  - [Quick configuration snippets](#quick-configuration-snippets)
    - [Flat ESLint config (eslint.config.js)](#flat-eslint-config-eslintconfigjs)
    - [Legacy .eslintrc.json](#legacy-eslintrcjson)
  - [Version](#version)
  - [Further Reading](#further-reading)

---

## Rule summary

- **Goal:** Ensure boolean-named identifiers (functions that return boolean, boolean-typed parameters, boolean properties, boolean variables) start with a recognizable boolean prefix.
- **Typical prefixes:** `is`, `has`, `can`, `should` (configurable).

---

## What the rule checks

1. **Function declarations** that return a boolean type (TypeScript-aware).
2. **Function expressions / arrow functions** assigned to variables (if they return boolean).
3. **Parameters** explicitly typed as boolean (including typed object parameter shapes and certain destructured parameter patterns).
4. **Object properties** whose values are boolean literals or functions returning boolean.
5. **Variables** explicitly typed boolean, variables initialized to boolean values, or variables initialized with function expressions/arrow functions that return boolean.

When a name is detected without an allowed prefix, the rule reports an error and includes a suggested corrected name (generated by prepending `is` and capitalizing the original name). The rule does **not** perform an automatic code fix — it reports and shows the suggestion in the message text.

---

## Options (all configurations)

The rule accepts a single options object (partial). Type definition:

```ts
type Options = [
    Partial<{
        allowedPrefixes: string[];
        checkFunctions: boolean;
        checkParameters: boolean;
        checkProperties: boolean;
        checkVariables: boolean;
        // ignore: regex string to skip identifiers matching the pattern
        // Example: "filter" will ignore any identifier containing 'filter'
        ignore: string;
    }>,
];
```

### Default options

```json
{
    "allowedPrefixes": [
        "is",
        "has",
        "can",
        "should",
        "will",
        "are",
        "were",
        "was"
    ],
    "checkFunctions": true,
    "checkParameters": true,
    "checkProperties": true,
    "checkVariables": true,
    "ignore": "filter"
}
```

### Option details

#### allowedPrefixes

- **Type:** `string[]`
- **Default:** `["is", "has", "can", "should", "will", "are", "were", "was"]`
- **Description:** List of prefixes that are considered valid for boolean identifiers. Matching is case-insensitive.
- **Example:** `["is","has","should","can","will"]`

#### ignore

- **Type:** `string`
- **Default:** `"filter"`
- **Description:** A regex string used to skip reporting identifiers whose names match this pattern (case-insensitive). Useful to avoid flagging known patterns like `somethingFilter` or `useFilter`.

#### checkFunctions

- **Type:** `boolean`
- **Default:** `true`
- **Description:** When `true`, the rule checks function declarations and variable-initialized functions that return boolean. If you set to `false`, functions will not be checked.

#### checkParameters

- **Type:** `boolean`
- **Default:** `true`
- **Description:** When `true`, the rule checks function parameters whose type is boolean (including certain typed object shapes and destructured patterns).

#### checkProperties

- **Type:** `boolean`
- **Default:** `true`
- **Description:** When `true`, the rule checks object properties that are booleans or functions returning booleans.

#### checkVariables

- **Type:** `boolean`
- **Default:** `true`
- **Description:** When `true`, the rule checks top-level and local variables (including variables initialized with boolean literals and variables initialized with functions returning boolean).

---

## Examples (by option)

> All examples use TypeScript where type resolution is necessary. For accurate function-return and variable-type checks, the rule relies on parserServices (TypeScript program).

### Default behavior

Incorrect:

```ts
function active(): boolean {
    return true;
}
const enabled = false;
const shouldProceed = true; // ok
const visible = () => true;
const config = { visible: true };
```

Correct:

```ts
function isActive(): boolean {
    return true;
}
const isEnabled = false;
const shouldProceed = true;
const isVisible = () => true;
const config = { isVisible: true };
```

### Custom prefixes

If you want `will` and `is` only:

```jsonc
{
    "rules": {
        "nima/boolean-naming-convention": [
            "error",
            { "allowedPrefixes": ["is", "will"] },
        ],
    },
}
```

Incorrect (with above config):

```ts
const hasAccess = true; // flagged (has not allowed)
```

Correct:

```ts
const isAccess = true; // or const willAccess = true;
```

(You’ll likely prefer semantic choices: consider choosing prefixes that read nicely in your codebase.)

### Disabling specific checks

If you want to check everything except function return names:

```jsonc
{
    "rules": {
        "nima/boolean-naming-convention": [
            "error",
            { "checkFunctions": false },
        ],
    },
}
```

Effect: a function that returns boolean can be named `active` without a flag, but parameters/properties/variables will still be checked.

#### Example

```ts
// With checkFunctions: false
function active(): boolean {
    return true;
} // not reported
const enabled = false; // still reported unless prefixed
```

### Destructured params & typed object patterns

The rule detects boolean members in typed parameter object shapes and some destructured patterns.

Incorrect:

```ts
function init(config: { enabled: boolean }) {
    /* ... */
} // flagged: enabled
function start({ running }: { running: boolean }) {
    /* ... */
} // flagged: running
function handle({ active }: { active: boolean }) {
    /* ... */
} // flagged: active
```

Correct:

```ts
function init(config: { isEnabled: boolean }) {
    /* ... */
}
function start({ isRunning }: { isRunning: boolean }) {
    /* ... */
}
function handle({ isActive }: { isActive: boolean }) {
    /* ... */
}
```

**Notes:** The rule checks:

- Parameters annotated with `TSTypeLiteral` (a parameter typed inline with a shape).
- Destructured parameters where property/identifier type annotations indicate boolean.

### Function-valued properties and variable initializers

The rule checks properties whose values are functions returning boolean, and variables initialized to functions returning boolean.

Incorrect:

```ts
const obj = {
    visible() {
        return true;
    }, // property name 'visible' flagged
};

const enabled = () => true; // flagged as a function-valued variable
```

Correct:

```ts
const obj = {
    isVisible() {
        return true;
    },
};

const isEnabled = () => true;
```

### Destructured variable declarators (typed)

If you destructure from an expression and TypeScript type information shows a property is boolean, the rule will check the destructured identifier.

Incorrect:

```ts
// given `options: { active: boolean }`
const { active } = options; // flagged if type info is present and active is boolean
```

Correct:

```ts
const { isActive } = options;
```

---

## Messages

When triggered, this rule emits one of the following messages (message templates include the suggested name):

- `BAD_FUNCTION_BOOLEAN_PREFIX`  
  Message: `NIMA: Function '{{name}}' returns a boolean, use a prefix like {{suggestion}}`

- `BAD_PARAMETER_BOOLEAN_PREFIX`  
  Message: `NIMA: Boolean parameter '{{name}}' should use a prefix like {{suggestion}}`

- `BAD_PROPERTY_BOOLEAN_PREFIX`  
  Message: `NIMA: Boolean property '{{name}}' should use a prefix like {{suggestion}}`

- `BAD_VARIABLE_BOOLEAN_PREFIX`  
  Message: `NIMA: Boolean variable '{{name}}' should use a prefix like {{suggestion}}`

**Example reported text:**

```text
NIMA: Function 'active' returns a boolean, use a prefix like isActive
```

---

## Implementation notes & requirements

- **Type information:** For reliable detection of return types and variable types this rule uses TypeScript type information via `context.sourceCode.parserServices.program` and `TypeChecker`. To enable this:
- Use `@typescript-eslint/parser` or ESLint Flat config with TypeScript program available.
- Ensure `parserOptions.project` points to your `tsconfig.json` (type-aware rules require a configured TypeScript program).
- **Graceful fallback:** Where type information is unavailable or an exception occurs, the rule falls back safely and will generally not report (avoids false positives).
- **Suggestion generation:** The suggestion is generated as `"is" + PascalCase(name)`. For example `visible` → `isVisible`.
- **No automatic fix:** This rule currently reports suggestions in messages but does not provide an ESLint autofix.

---

## Limitations & edge cases

- **Type resolution required:** Without parserServices and a valid TypeScript program, function-return and some variable type checks may not run (resulting in fewer reports).
- **False negatives possible:** If TypeScript cannot resolve a type or if code uses complex generic/conditional types, the rule may not detect the boolean type.
- **Name heuristics:** The rule checks prefix by string-start matching, case-insensitive. If your naming convention uses camelCase or prefixes elsewhere (like trailing suffixes), results may need custom prefixes or disabling the rule.
- **No auto-renaming:** The rule suggests a new identifier name in the message but does not modify your source code.

---

## Quick configuration snippets

### Flat ESLint config (eslint.config.js)

```js
import pluginNIMA from "eslint-plugin-nima";

export default [
    {
        plugins: { nima: pluginNIMA },
        rules: {
            "nima/boolean-naming-convention": [
                "error",
                {
                    allowedPrefixes: [
                        "is",
                        "has",
                        "can",
                        "should",
                        "will",
                        "are",
                        "were",
                        "was",
                    ],
                    checkFunctions: true,
                    checkParameters: true,
                    checkProperties: true,
                    checkVariables: true,
                },
            ],
        },
    },
];
```

### Legacy .eslintrc.json

```json
{
    "plugins": ["nima"],
    "rules": {
        "nima/boolean-naming-convention": [
            "error",
            {
                "allowedPrefixes": [
                    "is",
                    "has",
                    "can",
                    "should",
                    "will",
                    "are",
                    "were",
                    "was"
                ],
                "checkFunctions": true,
                "checkParameters": true,
                "checkProperties": true,
                "checkVariables": true
            }
        ]
    }
}
```

---

## Version

Introduced in `eslint-plugin-nima@1.0.0`.

---

## Further Reading

- [TypeScript Type Annotations](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html)
- [Naming Conventions in JavaScript](https://google.github.io/styleguide/jsguide.html#naming)
- [Clean Code: Meaningful Names](https://www.oreilly.com/library/view/clean-code-a/9780136083238/)
